Style Engine: block supports backend

ramonjd opened the pull request:
Tracking Issue: https://github.com/WordPress/gutenberg/issues/38167

## What?

A backend version of
- https://github.com/WordPress/gutenberg/pull/37978

~Experimenting with classname obfuscation.~ Reverted (see c[ommit history](https://github.com/WordPress/gutenberg/pull/39446/commits/eba46f468d8d3feeadca16ae2226c7d4872a0b6c#diff-df6c4ae64dee8330db2c09f5c60668cea91dfcbbad4e5143055c5c98e29ccbf5R106))

Works on block support styles (margin/padding for now) that are serverside rendered, e.g., `core/site-title` 

## Why?

To create a backend engine for style object -> css rule transformation to complement the frontend code: https://github.com/WordPress/gutenberg/pull/37978

## How?

`WP_Style_Engine_Gutenberg` has a `STYLE_DEFINITIONS_METADATA` dictionary that contains instructions on how to parse a Gutenberg style attributes in order to return CSS rules.

## Testing Instructions
Insert a few `core/site-title` blocks and add padding.

<details><summary>Editor code</summary>

 ```html
<!-- wp:site-title {"style":{"spacing":{"padding":{"top":"104px","right":"104px","bottom":"104px","left":"104px"},"margin":{"top":"72px","right":"72px","bottom":"72px","left":"72px"}},"color":{"background":"#f71c1c"}}} /-->

<!-- wp:site-title {"style":{"spacing":{"padding":{"top":"104px","right":"104px","bottom":"104px","left":"104px"},"margin":{"top":"286px","bottom":"286px","right":"286px","left":"286px"}},"color":{"background":"#ee8c8c"}}} /-->

<!-- wp:site-title {"style":{"spacing":{"padding":{"top":"104px","right":"104px","bottom":"148px","left":"104px"},"margin":{"top":"4px","right":"8.9vw","bottom":"4%","left":"4px"}},"color":{"background":"#f68888"}}} /-->

<!-- wp:site-title {"style":{"spacing":{"padding":{"bottom":"2vw","top":"1rem"},"margin":{"top":"3vh","right":"3vh","bottom":"3vh","left":"3vh"}},"color":{"background":"#896b6b"}}} /-->
```
</details>


Check the resulting output and CSS inline styles.

```html
<h1 style="background-color: #ee8c8c; padding-top: 104px; padding-right: 104px; padding-bottom: 104px; padding-left: 104px;  margin-top: 286px; margin-bottom: 286px; margin-right: 286px; margin-left: 286px; " class="has-background wp-block-site-title"><a href="http://localhost:4759" rel="home">a8c-wp-env</a></h1>
```
## Related PRs

- https://github.com/WordPress/gutenberg/pull/37978
- https://github.com/WordPress/gutenberg/pull/38974
- https://github.com/WordPress/gutenberg/pull/39374

ramonjd committed:
Initial and worst commit

ramonjd committed:
Adding a style definition box model utility to extract CSS rules from a Gutenberg style attributes object

ramonjd committed:
Adding inline style capabilities

ramonjd committed:
filter out empty values from the array

ramonjd committed:
Moved obfuscation comments to a php doc comment

Adding base test file

andrewserong commented on lib/style-engine/class-wp-style-engine-gutenberg.php:
```diff
@@ -0,0 +1,257 @@
+			return;
+		}
+
+		if ( ! empty( $options['obfuscate'] ) && true === $options['obfuscate'] ) {
```
I know you're still working on this, but just wanted to say I love this idea for where to inject the obfuscation — it neatly combines our two approaches by using the unique / semantic classname as the hash, so it's an easy switch between the two approaches ✨

ramonjd replied:
Thanks! Ultimately this was a side-experiment inside a side-experiment 😆 

It'd be nice to follow up on though.

Still, the minimum I suppose we could get from this PR is:

- replicating the current style engine functionality (padding and inline styles)
- tests (when I could be bothered writing them)

Not sure yet about moving the PHP file to the package. It's pretty tightly coupled to Gutenberg and there's no precedent for it yet. 

What do you reckon?

andrewserong replied:
> What do you reckon?

Conceptually, I think I liked the idea of keeping it in the package to begin with, but it's since become clearer that it'll get called from lots of other places and will be more tightly integrated into other Gutenberg files. I think I'd keep it in `lib/` for the time-being.

ramonjd commented on lib/style-engine/class-wp-style-engine-gutenberg.php:
```diff
@@ -0,0 +1,257 @@
+		if ( ! empty( $options['obfuscate'] ) && true === $options['obfuscate'] ) {
+			// Generate a consistent, obfuscated key by hashing a unique class based on the rules.
+			// Outputs something like .wp-my-key__002d308c.
+			$class = $key . '__' . hash( 'crc32', $class );
```
Maybe need some salting here to avoid collision?

ramonjd committed:
Removing experimental code dealing with obfuscation and class generation.

youknowriad commented on phpunit/style-engine/class-wp-style-engine-gutenberg-test.php:
```diff
@@ -0,0 +1,62 @@
+ */
+class WP_Style_Engine_Gutenberg_Test extends WP_UnitTestCase {
+	function test_returns_inline_styles_from_string() {
+		$style_engine   = WP_Style_Engine_Gutenberg::get_instance();
```
Does it need to be a class? Why not just a function, mostly curious about the reasoning here.

ramonjd replied:
Good question. It doesn't have to be a class. The reasons why I left it as a class are:

- because I copied the structure over from one of our initial experiments 😄 
- to namespace methods and encapsulate constants such as `BLOCK_STYLE_DEFINITIONS_METADATA` (similar to  `WP_Theme_JSON_Gutenberg`)
- portability across applications
- related to the last point, so we can be free to use generic method names such as `generate` to harmonize the backend and frontend APIs, as mentioned [below](https://github.com/WordPress/gutenberg/pull/39446#discussion_r832150344).
- while a more conceptual point, to lend the functionality a degree of authority and conspicuousness, that is, so folks recognize it as a unified collection of properties and behaviours.

andrewserong replied:
Good question, and good answers! 😁

> while a more conceptual point, to lend the functionality a degree of authority and conspicuousness, that is, so folks recognize it as a unified collection of properties and behaviours.

Yes, in PHP, because we're not using namespaces and can't encapsulate our functions in a separate module like we do in JS, I think it's better for us to gather together our style engine functions in a class rather than expose each of them in the global namespace. Plus, it's looking like we'll likely need to carry around some state in the style engine to deal with de-duping styles and / or deferring processing to a later part of the execution process, so starting with a class likely makes it easier for us to extend the style engine in follow-up PRs, too.

youknowriad replied:
Makes sense. I just wanted us to avoid the "stateful" trap if possible. When we add state to classes that are not meant to be classes, the problems start happening :)

youknowriad commented on phpunit/style-engine/class-wp-style-engine-gutenberg-test.php:
```diff
@@ -0,0 +1,62 @@
+				'padding' => '111px',
+			),
+		);
+		$padding_styles = $style_engine->get_inline_css_from_block_styles(
```
should we try to align the naming and arguments with the JS version? 

`generate( $style_object, $style_options )`

ramonjd replied:
Ah yeah, sounds like a good idea to be consistent across the libraries, even if to avoid confusion when reading both code bases 😄 

I try for verbosity since, in my opinion `generate` doesn't really tell much of a story unless we accept the implication that we're generating a string of raw styles. Documentation will mitigate this I think. 

Furthermore, if we introduce different return values such as formatted class rules or even classnames we could use the options array argument to drive them, or just create a new method.

ramonjd commented:
@youknowriad Thanks for your comments and help on this one. 

I wanted to also get your opinion on:

- whether we should move the style engine PHP file to the package now or later,
- and if and when we do move it, what's the best way to load it in Gutenberg? Just a require in `load.php`?

ramonjd committed:
Rename public method to `generate` to reflect current frontend API
Add optional options array
Explicitly tell `generate()` to return styles so we can open it up to further functionality
Update tests

aaronrobertshaw commented on lib/block-supports/spacing.php:
```diff
@@ -45,33 +45,46 @@ function gutenberg_apply_spacing_support( $block_type, $block_attributes ) {
-		} elseif ( null !== $margin_value ) {
-			$styles[] = sprintf( 'margin: %s;', $margin_value );
+		// As with padding above.
+		$margin_styles = $style_engine->generate(
+			$block_styles,
+			array( 'inline' => true )
+		);
```
Would this just duplicate the inline styles possibly already added via padding?

Without much context, it appears to me these generate calls should be asking for only a specific set of styles? In this case those for margin.

If the style engine doesn't check what's supported and the entire result is added here, we could end up with styles that aren't supported. For example, if `padding` is not supported but `margin` is and the style attribute contains `padding` values, wouldn't we erroneously get `padding` styles?

andrewserong replied:
It looks like the style engine in this PR is designed to only be called once per block, so it might become a problem when we roll out to other block supports, too.

I suppose we could either `unset` attributes that haven't opted in for support, or defer the support check to the style engine? I think I'd lean toward keeping the support checks within the block support, and just call `generate` once 🤔

ramonjd replied:
I committed a bunch of code without test thinking I was still in draft mode. Apologies and thanks for reporting. Getting the rest of the working code up soon...

andrewserong commented on phpunit/style-engine/class-wp-style-engine-gutenberg-test.php:
```diff
@@ -0,0 +1,111 @@
+					),
+				),
+				'options'         => array( 'inline' => true ),
+				'expected_output' => 'padding-top:42px;padding-left:2%;padding-bottom:44px;padding-right:5rem;',
```
Should this expected output contain both the padding and margin values?

andrewserong commented on lib/style-engine/class-wp-style-engine-gutenberg.php:
```diff
@@ -0,0 +1,157 @@
+ *
+ * @return array The class name for the added style.
+ */
+function gutenberg_get_style_engine_css_box_rules( $style_value, $style_property ) {
```
Just wondering if this should be a method of the `WP_Style_Engine` class? I see it's invoked via `call_user_func`, so I thought maybe we can use an array when setting `value_func` to pass in the class so that we call a static method, instead — or, do we prefer to keep some logic outside of that class?

ramonjd replied:
True, thanks for raising. 

I left it out because these formatting functions could really live anywhere and may perform some specific operations (like gap), so I didn't want to set a precedent for creating too many methods in the class. 

If you think this one in particular should live in the class, I'm all for it!

andrewserong replied:
I don't feel too strongly either way, but I suppose if we have it in the class, we have one fewer `gutenberg` prefixes to deal with? We can always rearrange in follow-ups if need be, though 🙂

ramonjd replied:
Fewer search and replaces at the start sounds like something I'm down for 😆 

Something like `'value_func'   => 'self::get_css_box_rules',` seems to work as well.

aaronrobertshaw commented on lib/style-engine/class-wp-style-engine-gutenberg.php:
```diff
@@ -0,0 +1,157 @@
+
+		$rules = array();
+
+		foreach ( $block_styles as $block_style => $block_style_properties ) {
+			if ( empty( $block_style ) || ! is_array( $block_style_properties ) ) {
+				continue;
+			}
+			$path        = array( $block_style, key( $block_style_properties ) );
+			$style_value = _wp_array_get( $block_styles, $path, null );
+
+			if ( empty( $style_value ) ) {
+				continue;
+			}
+
+			$rules = array_merge( $rules, $this->get_block_style_css_rules( $style_value, $path ) );
+		}
```
At a quick glance, this doesn't appear to loop through the individual style features e.g. `padding` and `margin`. 

For a block style attribute that includes both `padding` and `margin` within `spacing` we only process `padding`?

ramonjd replied:
Same excuse as [above](https://github.com/WordPress/gutenberg/pull/39446#discussion_r832820826) 🤣

andrewserong commented on lib/style-engine/class-wp-style-engine-gutenberg.php:
```diff
@@ -0,0 +1,157 @@
+			if ( empty( $block_style ) || ! is_array( $block_style_properties ) ) {
+				continue;
+			}
+			$path        = array( $block_style, key( $block_style_properties ) );
```
`$block_style_properties` is an array, so `key` here appears to be returning `margin` for me even if the `$block_style_properties` contains both `margin` and `padding` values.

Instead, do we need to loop over `$block_style_properties` in another for loop?

ramonjd replied:
Related to my complete brain short-circuit this morning. 🙇

aaronrobertshaw reviewed (commented):
Apologies if this isn't ready for a full review. I'm just a bit late to the party and trying to get my head around all the style engine work.

I've tried testing this as per the PR instructions however it doesn't work for me. 

Any margin styles added to the block are lost. In their place, I get duplicate inline padding styles. This occurred both with the example code and manually adding my own site title block.

| Trunk | This PR |
|---|---|
| <img width="292" alt="Screen Shot 2022-03-23 at 12 21 35 pm" src="https://user-images.githubusercontent.com/60436221/159614421-4d71a8f9-f656-4581-b83f-0f385f494d9b.png"> | <img width="211" alt="Screen Shot 2022-03-23 at 12 21 43 pm" src="https://user-images.githubusercontent.com/60436221/159614440-4862b520-d8ea-4e6b-a875-3538faa8426b.png"> |

I've added a couple of questions as inline comments. Feel free to ignore them if I've just missed something obvious.

andrewserong reviewed (commented):
Thanks for reworking this to be closer to the JS implementation, @ramonjd! I ran into similar issues to Aaron where the `margin` was rendered twice for the site title block:

<img width="646" alt="image" src="https://user-images.githubusercontent.com/14988353/159617716-f86d83a9-0c1a-44bb-a3a2-18e68199ee2e.png">

In terms of the style engine being called multiple times, or processing things more times than it should:

* Do we try to ensure that we're only calling `generate` once per block instance (across all block supports), and then defer as much as possible to the style engine? Or:
* Do we selectively call the style engine with only part of the `style` object. For instance, in the `spacing` support, perhaps we'd only call `generate` with an associative array containing `spacing` so that we don't accidentally process typography at the same time 🤔 

I think I'm leaning toward (at this stage) calling the style engine only once in the `spacing` support, and processing the style object a little bit before sending it to the style engine. What do you think?

ramonjd commented:
Thanks @aaronrobertshaw and @andrewserong for taking a look at this. 

Sorry for wasting your time with buggy code. I've pushed the code I should have pushed when I pushed this morning before I got distracted 🙏

ramonjd commented:
> Do we try to ensure that we're only calling generate once per block instance (across all block supports), and then defer as much as possible to the style engine? 

> Do we selectively call the style engine with only part of the style object. For instance, in the spacing support, perhaps we'd only call generate with an associative array containing spacing so that we don't accidentally process typography at the same time 🤔

> I think I'm leaning toward (at this stage) calling the style engine only once in the spacing support, and processing the style object a little bit before sending it to the style engine. What do you think?

I agree that we should only call it once. ✅  

It crossed my mind but I got bogged down thinking about how to determine the logic for `$has_padding_support` and `$has_margin_support`

I guess we could rebuild the style object before we pass it. Very rough (but working) example 😄 

```php
...
	$spacing_block_styles = array();

	if ( $has_padding_support ) {
		$padding_style_value = _wp_array_get( $block_styles, array( 'spacing', 'padding' ), null );
		if ( $padding_style_value ) {
			$spacing_block_styles['padding'] = $padding_style_value;
		}
	}

	if ( $has_margin_support ) {
		$margin_style_value = _wp_array_get( $block_styles, array( 'spacing', 'margin' ), null );
		if ( $margin_style_value ) {
			$spacing_block_styles['margin'] = $margin_style_value;
		}
	}

	$inline_styles = $style_engine->generate(
		array( 'spacing' => $spacing_block_styles ),
		array(
			'inline' => true,
		)
	);

...
	if ( ! empty( $inline_styles ) ) {
		$attributes['style'] = $inline_styles;
	}

	return $attributes;
```


Or we could have `generate()` accept more granular style rules, e.g., `{ padding: '12px' }` instead of the entire style object. 

In a later version I'd also like to be able to support passing a path like `array( 'spacing' )` which will then process _all_ the spacing styles only.


I was trying to keep it _kinda_ the same as the JS version, but maybe we don't need to be so strict about that? What do you reckon?

andrewserong commented:
Thanks for following up, it's a tricky balance to strike!

> I guess we could rebuild the style object before we pass it. Very rough (but working) example

I like that idea, I think it's worth exploring further. In terms of responsibilities, I think it makes sense that "block supports" owns whether or not something is supported, and the "style engine" owns generating the styles, so this kind of logic feels like a good direction to me.

> In a later version I'd also like to be able to support passing a path like array( 'spacing' ) which will then process all the spacing styles only.

Nice — yes, iteratively moving toward being able to call the style engine as few times as possible sounds good, and this sort of thing could mean calling once in `spacing` and then once in `typography`, etc.

> I was trying to keep it kinda the same as the JS version, but maybe we don't need to be so strict about that? What do you reckon?

I don't think we need to be _too_ strict, personally — as we implement things in PHP or in JS we'll encounter things that reveal changes we need to make in the other one. Here we've encountered two things, I think, that warrant going in a slightly different direction to what we expected if we need to:

* Needing to factor in whether or not the block has opted in to the support
* Dealing with padding _and_ margin (in JS we've only implemented one)

We'll then hopefully learn from the PHP implementation what's missing from the JS side and vice versa 😀

youknowriad commented:
> whether we should move the style engine PHP file to the package now or later,

I don't mind if it's done separately in another PR but I'd love if we can do soon to avoid falling in the trap of having two different JS/php implementations. Keeping them together kind of gives the necessary incentive that they should be implemented the same.

> and if and when we do move it, what's the best way to load it in Gutenberg? Just a require in load.php?

Yes, I think so. The difficulty is going to be more on the "naming" as we need to write the package as if it was consumed by core (so no gutenberg_ prefixes) but we also need to have a dedicated version in gutenberg that don't conflict with core (so gutenberg_ prefixes). The way we do this for the `block-library` functions is that we rewrite them on the fly in the webpack build.

ramonjd commented:
>  I think it makes sense that "block supports" owns whether or not something is supported,

When you put it like that it sounds so obvious! 😄 Good call.

> I don't mind if it's done separately in another PR but I'd love if we can do soon to avoid falling in the trap of having two different JS/php implementations. Keeping them together kind of gives the necessary incentive that they should be implemented the same.

Excellent. This is great to know. And thanks for the interesting tips. I'll take a closer look at the block-library. 🙇

ramonjd committed:
Building the spacing block styles before calling generate

andrewserong commented on lib/style-engine/class-wp-style-engine-gutenberg.php:
```diff
@@ -0,0 +1,166 @@
+	/**
+	 * Container for the main instance of the class.
+	 *
+	 * @since 5.5.0
```
Tiny nit: I was going to say, should this be `@since 6.0.0`, but because the entire class is new, I imagine we don't need to add `@since`? (I wasn't too sure what the practice is there in core 🤔)

ramonjd replied:
Copy paste job 😄

andrewserong commented on lib/style-engine/class-wp-style-engine-gutenberg.php:
```diff
@@ -0,0 +1,166 @@
+	 * @param string|array  $style_value A single raw Gutenberg style attributes value for a CSS property.
+	 * @param array<string> $path        An array of strings representing a path to the style value.
+	 *
+	 * @return array A CSS ruleset compatible with add_style().
```
Just a tiny change, since `add_style` was renamed to `generate`:

```suggestion
	 * @return array A CSS ruleset compatible with generate().
```

andrewserong commented on lib/style-engine/class-wp-style-engine-gutenberg.php:
```diff
@@ -0,0 +1,166 @@
+	 * @return string A CSS ruleset formatted to be placed in an HTML `style` attribute.
+	 */
+	public function generate( $block_styles, $options = array() ) {
+		$output = '';
```
Not a blocker at all for now, but I was wondering if we should move this `$output` and the `return $output` into the `if` block further down that checks for the `inline` option?

In follow-ups where we add in styles that aren't generated inline, we might use a different return type that could include both the generated styles and the classnames that refer to those styles. But we can play with all that in follow-ups, I just wanted to jot down the thought 😀

ramonjd replied:
> Not a blocker at all for now, but I was wondering if we should move this $output and the return $output into the if block further down that checks for the inline option?

I declare it at the top so that the return value is always consistent. The benefit, at least for now, is that you can concatenate the return value from `generate()` to an existing string, e.g., in `spacing.php`.

Also we return `$output` early in the first if block.

Would it still be better to  move them given that?

> In follow-ups where we add in styles that aren't generated inline, we might use a different return type that could include both the generated styles and the classnames that refer to those styles. But we can play with all that in follow-ups, I just wanted to jot down the thought 😀

Good point. Ultimately I guess we should return some sort of object with an `output` key or something.

andrewserong replied:
> Would it still be better to move them given that?

I think it's perfectly fine to keep it where it is for now, I was mostly just thinking out loud for what we might explore in follow-ups. I'll switch this conversation to resolved!

andrewserong commented on lib/style-engine/class-wp-style-engine-gutenberg.php:
```diff
@@ -0,0 +1,166 @@
+			// Generate inline style rules.
+			if ( isset( $options['inline'] ) && true === $options['inline'] ) {
+				foreach ( $rules as $rule => $value ) {
+					$output .= "{$rule}:{$value};";
```
Before outputting this... output, should we run it through [safecss_filter_attr](https://developer.wordpress.org/reference/functions/safecss_filter_attr/) a bit like we do [in the Search block](https://github.com/wordpress/gutenberg/blob/9c5549e0e55c9b165a7a57004fa102d7783a69eb/packages/block-library/src/search/index.php#L304)?

ramonjd replied:
We probably should. I suppose we don't do that in block supports right now because we assume that it's been cleaned along the editor pipeline. 

I looked at the [safe attributes list](https://github.com/WordPress/wordpress-develop/blob/5.9/src/wp-includes/kses.php#L2299) and noticed that `gap` isn't on there so we'll have to submit a core patch if we want to allow extra attributes.

For `margin-*` and `padding-*` we should be golden though - I'll add it in to be safe.

aaronrobertshaw replied:
@ramonjd I believe you can add to that safe attributes list via the [`gutenberg_safe_style_attrs `](https://github.com/WordPress/gutenberg/blob/596afccfe1396584f11e7c5b7cc22bf1b24e4ced/lib/compat.php#L19-L29) in `lib/compat.php`. They will then be ported to core in due time.

andrewserong reviewed (commented):
Thanks for fixing this up @ramonjd, I think this is looking very close to being in a shape we could land! Just added a couple more tiny comments, but I really like how it looks — now that the block supports is just responsible for checking which supports are available and then calling out to the style engine, I think it shows a clearer relationship between how the two APIs will work 👍

Tested with the Site Title block:

![image](https://user-images.githubusercontent.com/14988353/159858035-4bb7d931-3230-4fa7-a8c1-36a8b61c2de9.png)

✅ Renders padding on its own
✅ Renders margin on its own
✅ Renders padding and margin together
✅ Skips padding and / or margin if you disable either block support in `block.json`
✅ Other block supports (e.g. typography) are unaffected by this addition

It sounds like we can look into re-arranging where we put the file in a follow-up (so we can then dig into the webpack build process and see about removing the Gutenberg prefix / suffixes) 👍

It'd be good to get another couple pairs of eyes on this as well, so that we can confidence check with some folks who aren't as familiar with what we've been trying to do, too 😀

Thanks again for all the follow-ups!

ramonjd committed:
Move box style function into the class
Update comments, fix typos
Filter CSS output

ramonjd committed:
Try:
- move style engine into `/packages/style-engine`
- create webpack copy pattern for packages with classes
- switched class methods to static so we don't have to get the instance

ramonjd commented:
> The difficulty is going to be more on the "naming" as we need to write the package as if it was consumed by core (so no gutenberg_ prefixes) but we also need to have a dedicated version in gutenberg that don't conflict with core (so gutenberg_ prefixes). The way we do this for the block-library functions is that we rewrite them on the fly in the webpack build.

I'm not sure if I'm on the right track but I decided to try out copying the PHP file from `packages/style-engine` over to `/lib` during the package build in 3a9ef2b45c2dbc1311ed0bf4d48c36eb5567d6d3

It's a commit that we can roll back and patch over to a new PR (in fact, I think we probably should since it might be nice to add documentation), I just wanted to test the approach on this branch to see if the tests pass 😄 

Also I wanted to get things ready so we can merge the copy packages PR soon after this one. 

Once the tests pass I'll create a patch from 3a9ef2b45c2dbc1311ed0bf4d48c36eb5567d6d3 and roll it back. 🍺

youknowriad commented on lib/block-supports/spacing.php:
```diff
@@ -51,11 +51,10 @@ function gutenberg_apply_spacing_support( $block_type, $block_attributes ) {
 	$spacing_block_styles['padding'] = $has_padding_support ? _wp_array_get( $block_styles, array( 'spacing', 'padding' ), null ) : null;
 	$spacing_block_styles['margin']  = $has_margin_support ? _wp_array_get( $block_styles, array( 'spacing', 'margin' ), null ) : null;
-	$inline_styles                   = $style_engine->generate(
+	$inline_styles                   = WP_Style_Engine_Gutenberg::generate(
```
This raises an interesting question that doesn't need to be answered now though :P 

Basically here we're using the "gutenberg" name and not the core name. This works for now because we're back porting the block supports code manually, but we were considering having automatic backports (like block-library and parser) for block supports changes. So when we do that, we'll have to find a way to use the core name here and also rename it to the gutenberg name on the build step.

ramonjd replied:
> So when we do that, we'll have to find a way to use the core name here and also rename it to the gutenberg name on the build step.

Yeah, I was thinking we could have a single function somewhere that returns `WP_Style_Engine_Gutenberg::get_instance()` so we'd only have one place to update it. 🤷

ramonjd committed:
Roll back previous test commit. There were build errors that might best be resolved in another PR.

ramonjd commented:
> I'm not sure if I'm on the right track but I decided to try out copying the PHP file from packages/style-engine over to /lib during the package build in https://github.com/WordPress/gutenberg/commit/3a9ef2b45c2dbc1311ed0bf4d48c36eb5567d6d3

I rolled it back already. Although the build step is part of the PHP tests (and I assumed the webpack copy), they still fail for some reason. I'll throw up a separate PR for moving the style engine PHP file to the package folder so we can get some expert 👀 on it. 👍

andrewserong commented on phpunit/style-engine/class-wp-style-engine-gutenberg-test.php:
```diff
@@ -0,0 +1,156 @@
+ * Tests the Style Engine class and associated functionality.
+ *
+ * @package    Gutenberg
+ * @subpackage block-library
```
Tiniest of nits: once we've moved the style engine class in a follow-up, this'll be:

```suggestion
 * @subpackage style-engine
```

(We can address that in the follow-up PR if you'd like)

ramonjd replied:
Ctrl-C Ctrl-V 💥

andrewserong commented on lib/load.php:
```diff
@@ -121,6 +121,11 @@ function gutenberg_is_experiment_enabled( $name ) {
 require __DIR__ . '/global-styles.php';
 require __DIR__ . '/pwa.php';
 
+// TODO: Before this PR merges, move this to be a part of the style engine package.
```
Shall we update this comment slightly?

```suggestion
// TODO: Move this to be loaded from the style engine package, via the build directory.
```

andrewserong reviewed (approved):
Thanks for all the follow-ups, @ramonjd!

Just gave this a re-test and it's still working nicely for me, and I tested the tests locally to make sure that they all break in expected ways if we fiddle with the expected data 👍

This LGTM as a place for us to start now, particularly with the intended follow-up PR to look at the build process and switching out of the `Gutenberg` suffix in the classname.

Just left a couple of tiny comments about comments, but nothing major. Would it be good to get a second approval since I'm maybe _slightly_ biased toward the implementation? 😁

ramonjd committed:
Apply suggestions from code review - update comments

Co-authored-by: Andrew Serong <14988353+andrewserong@users.noreply.github.com>

youknowriad reviewed (approved):
I didn't check deeply but this looks good to me.

ramonjd commented:
Given that I've sneaked `margin` support into this one, I've added it to the JS version over [here](https://github.com/WordPress/gutenberg/pull/39790) for consistency.

glendaviesnz commented:
This all worked as advertised for me:

✅ Files as copied to /build folder and renamed and `Gutenberg` appended to class name
✅ Unit tests passed
✅ Styles were added to the sampe `h1` blocks on frontend
